draft-ietf-avtcore-rtp-over-quic.md

```diff
@@ -1110,6 +1110,32 @@ to determine if 0-RTT data is permissible.
 > the same connection, replayed media packets would be discarded as duplicates by
 > the receiver.
 
+## Coalescing RTP packets in single QUIC packet
+
+
+Applications usually cannot control how the QUIC stack maps application data to
+QUIC packets. While the application controls the content of QUIC DATAGRAM
+frames, stream frames may contain any amount of data. QUIC packets can include
+multiple STREAM frames from different streams and one or more DATAGRAM frames.
+Applications don't necessarily control how the QUIC stack maps STREAM and
+DATAGRAM frames to QUIC packets {{Section 13 of ?RFC9000}} and {{Section 5 of
+?RFC9308}}.
+
+If there is not enough data, QUIC stacks might wait before sending out packets
+to optimize for bandwidth utilization instead of latency, while real-time
+applications usually prefer to optimize for latency rather than bandwidth
+utilization.
+
+When there are a lot of small DATAGRAM frames (e.g., an audio stream) and a lot
+of large DATAGRAM (e.g., a video stream), it may be a good idea to make sure the
+audio frames can be included in a QUIC packet that also carries video frames
+(i.e., the video frames don't fill the whole QUIC packet). Otherwise, the QUIC
+stack may have to send additional small packets only carrying single audio
+frames, which would waste some bandwidth.
```

Comment:
Just to make sure I understand this, is the recommendation that the application queue any related small media objects with an earlier RTP timestamp than the next large media object, before they queue that large media object?

I'm trying to figure out how the application can make sure that the audio frames can be included in a QUIC packet that also carries video frames. I think "sending small objects first" might accomplish that, but whatever we say, we need to give advice that an application developer can implement. 

This only matters if we don't have lots of large objects, right?

Reply:
I think any recommendation we can make depends on the level of control the application has over the QUIC stack. Queuing small objects before large ones might work but may require waiting for small objects in the application and possibly delaying large objects. Or delaying small objects to fill the packet and avoid sending small packets. There are some contradictory goals because we'd like to send audio packets as soon as possible but also try to avoid sending small packets, thus having to wait until a packet is _large enough_. It may require some research to give good recommendations here.

Reply:
That all makes perfect sense to me. The question is, how the QUIC implementation might adjust its delay timer so that the right thing happens. 

But what I was suggesting, was that if you queue up all the small frames with a timestamp earlier than the next large frame, that should minimize the chance that the QUIC implementation will emit a bunch of small packets. But this is guessing at what the QUIC implementation might do, in the absence of any defined behavior (beyond "I might wait for a short time before sending the packet"), and in the absence of any way to tell the QUIC implementation "if you're waiting a short time for more frames, this stuff is audio, so please set your timer appropriately". 

Maybe THIS is something the "rtp-mux-quic" ALPN could tell the QUIC implementation? 

I love heuristics about how to deal with someone else's heuristics. :grimacing: